module/zfs/arc.c

```diff
@@ -5611,6 +5611,53 @@ arc_access(arc_buf_hdr_t *hdr, kmutex_t *hash_lock)
 	}
 }
 
+/*
+ * This routine is called by dbuf_hold() to update the arc_access()
+ * information which otherwise would be skipped for dbufs in the cache.
+ */
+void
+arc_buf_access(arc_buf_t *buf)
+{
+	arc_buf_hdr_t *hdr;
+	kmutex_t *hash_lock;
+
+	mutex_enter(&buf->b_evict_lock);
+	hdr = buf->b_hdr;
+
+	if (hdr->b_l1hdr.b_state == arc_anon || HDR_EMPTY(hdr)) {
+		mutex_exit(&buf->b_evict_lock);
+		return;
+	}
+
+	/*
+	 * Because the hash_lock is obtained from the arc_buf_hdr_t it is
+	 * possible the arc_buf_hdr_t may be released before the hash_lock
+	 * can be acquired.  Detect this case by verifying the arc_buf_hdr_t
+	 * is still hashed after taking the lock.
+	 */
+	hash_lock = HDR_LOCK(buf->b_hdr);
+	mutex_enter(hash_lock);
+
+	if (!HDR_IN_HASH_TABLE(hdr)) {
+		mutex_exit(hash_lock);
+		mutex_exit(&buf->b_evict_lock);
+		return;
+	}
+
+	mutex_exit(&buf->b_evict_lock);
```

Comment:
lock b_evict_lock
lock hash_lock
unlock b_evict_lock
unlock hash_lock

I assume this is deadlock safe?

Reply:
Yes it should be, the same ordering is used throughout the ARC.